File: Tests/Categories/TestMessageUI.m
@import MessageUI;
@import UIKit;
#import "UIViewController+AnyPromise.h"
@import XCTest;


#if 0  // not possible to test these in the simulator

@implementation Test_MessageUI_ObjC: XCTestCase {
    UIViewController *rootvc;
}

- (void)setUp {
    rootvc = [UIApplication sharedApplication].keyWindow.rootViewController = [UIViewController new];
}

- (void)tearDown {
    rootvc = [UIApplication sharedApplication].keyWindow.rootViewController = nil;
}

- (void)testCanCancelMailComposeViewController {
    id ex = [self expectationWithDescription:@""];

    MFMailComposeViewController *mf = [MFMailComposeViewController new];

    [mf setToRecipients:@[@"mxcl@me.com"]];

    [rootvc promiseViewController:mf animated:NO completion:^{
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            UIBarButtonItem *button = [mf.viewControllers[0] navigationItem].leftBarButtonItem;

            ((void (*)(id, SEL))[button.target methodForSelector:button.action])(button.target, button.action);
        });
    }].catchWithPolicy(PMKCatchPolicyAllErrors, ^{
        [ex fulfill];
    });

    [self waitForExpectationsWithTimeout:10 handler:nil];
}

- (void)testCanCancelMessageComposeViewController {
    id ex = [self expectationWithDescription:@""];

    MFMessageComposeViewController *mf = [MFMessageComposeViewController new];

    [rootvc promiseViewController:mf animated:NO completion:^{
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.5 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
            UIBarButtonItem *button = [mf.viewControllers[0] navigationItem].leftBarButtonItem;

            ((void (*)(id, SEL))[button.target methodForSelector:button.action])(button.target, button.action);
        });
    }].catchWithPolicy(PMKCatchPolicyAllErrors, ^{
        [ex fulfill];
    });

    [self waitForExpectationsWithTimeout:10 handler:nil];
}

@end

#endif

File: Tests/Categories/TestMessageUI.swift
import MessageUI
import PromiseKit
import UIKit
import XCTest

#if false  // not possible to test these in the simulator

class Test_MessageUI_Swift: UIKitTestCase {
    func test_can_cancel_mail_composer() {
        let ex1 = expectationWithDescription("")
        let ex2 = expectationWithDescription("")
        var order = false

        let mailer = MFMailComposeViewController()
        mailer.setToRecipients(["mxcl@me.com"])

        let promise = rootvc.promiseViewController(mailer, animated: false, completion: {
            after(0.25).then { _ -> Void in
                XCTAssertFalse(order)
                let button = mailer.viewControllers[0].navigationItem.leftBarButtonItem!
                UIControl().sendAction(button.action, to: button.target, forEvent: nil)
                ex1.fulfill()
            }
        })
        promise.error { _ -> Void in
            XCTFail()
        }
        promise.error(policy: .AllErrors) { _ -> Void in
            // seems necessary to give vc stack a bit of time
            after(0.5).then(ex2.fulfill)
            order = true
        }
        waitForExpectationsWithTimeout(10, handler: nil)

        XCTAssertNil(rootvc.presentedViewController)
    }

    func test_can_cancel_message_composer() {
        let ex1 = expectationWithDescription("")
        let ex2 = expectationWithDescription("")
        var order = false

        let messager = MFMessageComposeViewController()

        let promise = rootvc.promiseViewController(messager, animated: false, completion: {
            after(0.25).then { _ -> Void in
                XCTAssertFalse(order)

                let button = messager.viewControllers[0].navigationItem.leftBarButtonItem!
                UIControl().sendAction(button.action, to: button.target, forEvent: nil)
                ex1.fulfill()
            }
        })

        promise.error { _ -> Void in
            XCTFail()
        }
        promise.error(policy: .AllErrors) { _ -> Void in
            // seems necessary to give vc stack a bit of time
            after(0.5).then(ex2.fulfill)
            order = true
        }
        waitForExpectationsWithTimeout(10, handler: nil)

        XCTAssertNil(rootvc.presentedViewController)
    }
}

#endif
